File: multipaz-compose/src/commonMain/kotlin/org/multipaz/compose/items/FloatingItemHeadingAndContent.kt
package org.multipaz.compose.items

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

/**
 * An item showing a field heading with custom content below it.
 *
 * This is intended for key-value inspector and detail views.
 *
 * @param heading field label shown at the top in semi-bold.
 * @param content custom content to display below the heading.
 * @param modifier a [Modifier].
 * @param showChevron whether to show a right chevron icon on the right side.
 * @param image optional image, shown to the left of the text.
 * @param trailingContent optional trailing content.
 */
@Composable
fun FloatingItemHeadingAndContent(
    heading: String,
    content: @Composable () -> Unit,
    modifier: Modifier = Modifier,
    showChevron: Boolean = false,
    image: @Composable () -> Unit = {},
    trailingContent: @Composable () -> Unit = {},
) {
    FloatingItemContainer(
        modifier = modifier,
        showChevron = showChevron,
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(16.dp, alignment = Alignment.Start),
            verticalAlignment = Alignment.CenterVertically
        ) {
            image()
            Column(
                modifier = Modifier.weight(1.0f),
                verticalArrangement = Arrangement.spacedBy(2.dp)
            ) {
                Text(
                    text = heading,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onSurface
                )
                content()
            }
            trailingContent()
        }
    }
}

File: multipaz-swiftui/src/iosMain/swift/FloatingItemHeadingAndContent.swift
import SwiftUI

/// A list item view that displays a bold heading with content below it,
/// an optional leading image, and an optional trailing view.
public struct FloatingItemHeadingAndContent<ContentView: View, ImageView: View, TrailingView: View>: View {

    /// The heading string displayed in a bold font above the primary text.
    public var heading: String

    /// Whether to show a right chevron icon on the right side.
    public var showChevron: Bool

    /// A view builder that creates the content.
    @ViewBuilder public var content: () -> ContentView

    /// A view builder that creates the leading image or icon.
    @ViewBuilder public var image: () -> ImageView

    /// A view builder that creates the trailing content, pushed to the rightmost edge.
    @ViewBuilder public var trailingContent: () -> TrailingView

    /// Creates a new floating item with a heading and content.
    ///
    /// - Parameters:
    ///   - heading: The bold heading to display.
    ///   - showChevron: Whether to show a right chevron icon on the right side. Defaults to `false`.
    ///   - content: A view builder that provides the content.
    ///   - image: A view builder that provides a leading image or icon. Defaults to an `EmptyView`.
    ///   - trailingContent: A view builder that provides a trailing view. Defaults to an `EmptyView`.
    public init(
        heading: String,
        showChevron: Bool = false,
        @ViewBuilder content: @escaping () -> ContentView,
        @ViewBuilder image: @escaping () -> ImageView = { EmptyView() },
        @ViewBuilder trailingContent: @escaping () -> TrailingView = { EmptyView() }
    ) {
        self.heading = heading
        self.showChevron = showChevron
        self.content = content
        self.image = image
        self.trailingContent = trailingContent
    }

    public var body: some View {
        FloatingItemContainer(showChevron: showChevron) {
            HStack(alignment: .center, spacing: 16) {
                image()

                VStack(alignment: .leading, spacing: 4) {
                    Text(heading)
                        .font(.system(size: 15))
                        .fontWeight(.bold)

                    content()
                }

                // Pushes the trailing content to the rightmost edge
                Spacer(minLength: 0)

                trailingContent()
            }
        }
    }
}
